packages/contracts/contracts/voting/multisig/Multisig.sol

```diff
@@ -0,0 +1,398 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import {ERC165Upgradeable} from "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import {_uncheckedAdd, _uncheckedSub} from "../../utils/UncheckedMath.sol";
+import {IDAO} from "../../core/IDAO.sol";
+import {GovernancePluginUUPSUpgradeable} from "../../core/plugin/GovernancePluginUUPSUpgradeable.sol";
+import {IMajorityVoting} from "../majority/IMajorityVoting.sol";
+import {Addresslist} from "../addresslist/Addresslist.sol";
+
+/// @title Multisig
+/// @author Aragon Association - 2022.
+/// @notice The on-chain multisig governance plugin in which a proposal passes if X out of Y approvals are met.
+/// @dev This contract inherits from `MajorityVotingBase` and implements the `IMajorityVoting` interface.
+contract Multisig is
+    Initializable,
+    ERC165Upgradeable,
+    Addresslist,
+    GovernancePluginUUPSUpgradeable
+{
+    /// @notice A container for proposal-related information.
+    /// @param executed Wheter the proposal is executed or not.
+    /// @param parameters The proposal-specific approve settings at the time of the proposal creation.
+    /// @param tally The approve tally of the proposal.
+    /// @param approvers The approves casted by the approvers.
+    /// @param actions The actions to be executed when the proposal passes.
+    struct Proposal {
+        bool open;
+        bool executed;
+        ProposalParameters parameters;
+        Tally tally;
+        mapping(address => bool) approvers;
+        IDAO.Action[] actions;
+    }
+
+    /// @notice A container for the proposal parameters.
+    /// @param minApprovals The number of approvals required.
+    /// @param snapshotBlock The number of the block prior to the proposal creation.
+    struct ProposalParameters {
+        uint256 minApprovals;
+        uint64 snapshotBlock;
+    }
+
+    /// @notice A container for the proposal tally.
+    /// @param approvals The number of approvals casted.
+    /// @param addresslistLength The length of the .
+    struct Tally {
+        uint256 approvals;
+        uint256 addresslistLength;
+    }
+
+    /// @notice The [ERC-165](https://eips.ethereum.org/EIPS/eip-165) interface ID of the contract.
+    bytes4 internal constant MULTISIG_INTERFACE_ID =
+        this.addAddresses.selector ^
+            this.removeAddresses.selector ^
+            this.isListed.selector ^
+            this.addresslistLength.selector ^
+            this.initialize.selector;
+
+    /// @notice The ID of the permission required to call the `addAddresses` and `removeAddresses` functions.
+    bytes32 public constant UPDATE_MULTISIG_SETTINGS_PERMISSION_ID =
```

Comment:
`UPDATE_PLUGIN_SETTINGS_PERMISSION` maybe?

Reply:
I would strongly suggest to writing specific permission identifiers instead of having one global permission for everything.
In the end, seeing a long list of permission, you don't know anymore which settings are doing what.

Reply:
Created APP-1447 for this

Reply:
Resolved as this will be tackled in APP-1447